File: sdk/python/src/tinyverse/types/messaging.py
from __future__ import annotations

from dataclasses import dataclass
from typing import Literal, Optional


EnvelopeType = Literal["CIPHERTEXT", "PREKEY_BUNDLE"]

ContentHint = Literal["DEFAULT", "RESENDABLE", "IMPLICIT"]


@dataclass
class SignalMetadata:
    ephemeral_key: Optional[str] = None
    signed_pre_key_id: Optional[str] = None
    one_time_pre_key_id: Optional[str] = None
    ratchet_key: Optional[str] = None
    message_number: Optional[int] = None
    previous_chain_length: Optional[int] = None
    sender_key_id: Optional[str] = None
    sender_key_iteration: Optional[int] = None
    rotation_required: Optional[bool] = None
    rotation_id: Optional[str] = None
    rotation_epoch: Optional[int] = None
    removed_agent_id: Optional[str] = None


@dataclass
class MessageEnvelope:
    id: str
    from_: str
    to: str
    timestamp: str
    device_id: int
    type: EnvelopeType
    body: str
    content_hint: Optional[ContentHint] = None
    signal: Optional[SignalMetadata] = None


@dataclass
class MessageStats:
    agent_id: str
    messages_sent: int
    unique_recipients: int


@dataclass
class MessageDeliveryReceipt:
    message_id: str
    from_: str
    to: str
    acknowledged_by: str
    acknowledged_at: str


@dataclass
class SignedKey:
    key_id: str
    public_key: str
    signature: Optional[str] = None


@dataclass
class KeyBundle:
    agent_id: str
    identity_key: str
    signed_pre_key: SignedKey
    updated_at: str
    one_time_pre_key: Optional[SignedKey] = None


@dataclass
class KeyHealth:
    agent_id: str
    one_time_pre_key_count: int
    low_one_time_pre_keys: bool
    updated_at: str
    signed_pre_key_key_id: Optional[str] = None
    signed_pre_key_updated_at: Optional[str] = None


@dataclass
class PreKeysRequest:
    pre_keys: list[SignedKey]
    identity_key: Optional[str] = None


@dataclass
class SignedPreKeyRequest:
    signed_pre_key: SignedKey
    identity_key: Optional[str] = None

File: sdk/rust/src/types/messaging.rs
#[allow(unused_imports)]
use super::*;
use serde::{Deserialize, Serialize}; // sibling types share a flat namespace, like the TS barrel

/// `"CIPHERTEXT" | "PREKEY_BUNDLE"` in the TS SDK.
pub type EnvelopeType = String;

/// `"DEFAULT" | "RESENDABLE" | "IMPLICIT"` in the TS SDK.
pub type ContentHint = String;

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SignalMetadata {
    #[serde(skip_serializing_if = "Option::is_none", default)]
    pub ephemeral_key: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none", default)]
    pub signed_pre_key_id: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none", default)]
    pub one_time_pre_key_id: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none", default)]
    pub ratchet_key: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none", default)]
    pub message_number: Option<i64>,
    #[serde(skip_serializing_if = "Option::is_none", default)]
    pub previous_chain_length: Option<i64>,
    #[serde(skip_serializing_if = "Option::is_none", default)]
    pub sender_key_id: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none", default)]
    pub sender_key_iteration: Option<i64>,
    #[serde(skip_serializing_if = "Option::is_none", default)]
    pub rotation_required: Option<bool>,
    #[serde(skip_serializing_if = "Option::is_none", default)]
    pub rotation_id: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none", default)]
    pub rotation_epoch: Option<i64>,
    #[serde(skip_serializing_if = "Option::is_none", default)]
    pub removed_agent_id: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct MessageEnvelope {
    pub id: String,
    pub from: String,
    pub to: String,
    pub timestamp: String,
    pub device_id: i64,
    #[serde(rename = "type")]
    pub envelope_type: EnvelopeType,
    pub body: String,
    #[serde(skip_serializing_if = "Option::is_none", default)]
    pub content_hint: Option<ContentHint>,
    #[serde(skip_serializing_if = "Option::is_none", default)]
    pub signal: Option<SignalMetadata>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct MessageStats {
    pub agent_id: String,
    pub messages_sent: i64,
    pub unique_recipients: i64,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct MessageDeliveryReceipt {
    pub message_id: String,
    pub from: String,
    pub to: String,
    pub acknowledged_by: String,
    pub acknowledged_at: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SignedKey {
    pub key_id: String,
    pub public_key: String,
    #[serde(skip_serializing_if = "Option::is_none", default)]
    pub signature: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct KeyBundle {
    pub agent_id: String,
    pub identity_key: String,
    pub signed_pre_key: SignedKey,
    #[serde(skip_serializing_if = "Option::is_none", default)]
    pub one_time_pre_key: Option<SignedKey>,
    pub updated_at: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct KeyHealth {
    pub agent_id: String,
    pub one_time_pre_key_count: i64,
    pub low_one_time_pre_keys: bool,
    #[serde(skip_serializing_if = "Option::is_none", default)]
    pub recommended_pre_key_refill: Option<i64>,
    #[serde(skip_serializing_if = "Option::is_none", default)]
    pub signed_pre_key_key_id: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none", default)]
    pub signed_pre_key_updated_at: Option<String>,
    pub updated_at: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct PreKeysRequest {
    #[serde(skip_serializing_if = "Option::is_none", default)]
    pub identity_key: Option<String>,
    pub pre_keys: Vec<SignedKey>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SignedPreKeyRequest {
    #[serde(skip_serializing_if = "Option::is_none", default)]
    pub identity_key: Option<String>,
    pub signed_pre_key: SignedKey,
}
